File: 04_mergeIntervals/intersection.py
def merge(arr_a, arr_b):
  result = []
  p1, p2, start, end = 0, 0, 0, 1

  while p1 < len(arr_a) and p2 < len(arr_b):
    b_overlaps_a = arr_a[p1][start] <= arr_b[p2][start] and arr_a[p1][end] >= arr_b[p2][start]
    a_overlaps_b = arr_b[p2][start] <= arr_a[p1][start] and arr_b[p2][end] >= arr_a[p1][start]

    if (a_overlaps_b or b_overlaps_a):
      result.append([max(arr_a[p1][start], arr_b[p2][start]), 
        min(arr_a[p1][end], arr_b[p2][end])])

    if arr_a[p1][end] < arr_b[p2][end]:
      p1 += 1
    else:
      p2 += 1

  return result


def main():
  print("Intervals Intersection: " + str(merge([[1, 3], [5, 6], [7, 9]], [[2, 3], [5, 7]])))
  print("Intervals Intersection: " + str(merge([[1, 3], [5, 7], [9, 12]], [[5, 10]])))


main()

File: 04_mergeIntervals/intersection.cpp
using namespace std;

#include <iostream>
#include <vector>

class Interval {
 public:
  int start = 0;
  int end = 0;

  Interval(int start, int end) {
    this->start = start;
    this->end = end;
  }
};

class IntervalsIntersection {
 public:
  static vector<Interval> merge(const vector<Interval> &arr1, const vector<Interval> &arr2) {
    vector<Interval> result;
    int p1=0;
    int p2=0;
    while(p1<arr1.size()&&p2<arr2.size()) {
      if((arr1[p1].start<=arr2[p2].start && arr1[p1].end>=arr2[p2].start) ||
        (arr2[p2].start<=arr1[p1].start && arr2[p2].end>=arr1[p1].start) ) {
        int start=max(arr1[p1].start,arr2[p2].start);
        int end=min(arr1[p1].end,arr2[p2].end);
        result.push_back({start,end});
      }
      if(arr1[p1].end<arr2[p2].end) {
        p1++;
      } else {
        p2++;
      }
    }
    return result;
  }
};

int main(int argc, char *argv[]) {
  vector<Interval> input1 = {{1, 3}, {5, 6}, {7, 9}};
  vector<Interval> input2 = {{2, 3}, {5, 7}};
  vector<Interval> result = IntervalsIntersection::merge(input1, input2);
  cout << "Intervals Intersection: ";
  for (auto interval : result) {
    cout << "[" << interval.start << "," << interval.end << "] ";
  }
  cout << endl;

  input1 = {{1, 3}, {5, 7}, {9, 12}};
  input2 = {{5, 10}};
  result = IntervalsIntersection::merge(input1, input2);
  cout << "Intervals Intersection: ";
  for (auto interval : result) {
    cout << "[" << interval.start << "," << interval.end << "] ";
  }
}
